### PR TITLE
Tobin gh1951 finish explicit vus

### PR DESCRIPTION
**Pull request by tobine**
This is the rest of the explicit VUs for #1951 along with tests. Still need to handle secondary CB cases.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -6362,14 +6362,96 @@ static bool ValidateRenderPassPipelineBarriers(layer_data *device_data, const ch
...
+    VK_PIPELINE_STAGE_COMMAND_PROCESS_BIT_NVX,
+    // VK_ACCESS_COMMAND_PROCESS_WRITE_BIT_NVX = 18
+    VK_PIPELINE_STAGE_COMMAND_PROCESS_BIT_NVX,
+};
+
+// Verify that all bits of access_mask are supported by the src_stage_mask
+static bool ValidateAccessMaskPipelineStage(VkAccessFlags access_mask, VkPipelineStageFlags stage_mask) {
+    // Early out if all commands set, or access_mask NULL
+    if (stage_mask & VK_PIPELINE_STAGE_ALL_COMMANDS_BIT) return true;
+    if (0 == access_mask) return false;
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: This is incorrect. It's valid and useful to have a zero access mask (on the src side!).

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -6362,14 +6362,96 @@ static bool ValidateRenderPassPipelineBarriers(layer_data *device_data, const ch
...
+    stage_mask = ExpandPipelineStageFlags(stage_mask);
+    int index = 0;
+    bool supported = true;
+    // for each of the set bits in access_mask, make sure that supporting stage mask bit(s) are set
+    while (access_mask) {
+        index = __builtin_ctz(access_mask);
+        assert(index >= 0);
+        supported &= AccessMaskToPipeStage[index] & stage_mask;
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: May as well just bail here if unsupported.

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -10557,8 +10563,7 @@ TEST_F(VkLayerTest, InvalidBarriers) {
     VkCommandBufferObj bad_command_buffer(m_device, &command_pool);
 
     bad_command_buffer.begin();
-    buf_barrier.offset = 0;
-    buf_barrier.size = VK_WHOLE_SIZE;
+    buf_barrier.srcAccessMask = VK_ACCESS_SHADER_READ_BIT;
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Read bits on src side are (benign) nonsense. Did you mean 0?

File: tests/layer_validation_tests.cpp [hunk 4]
```diff
@@ -10254,14 +10254,14 @@ TEST_F(VkLayerTest, InvalidBarriers) {
     img_barrier.subresourceRange.levelCount = 1;
     m_commandBuffer->begin();
     vkCmdPipelineBarrier(m_commandBuffer->handle(), VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
-                         VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT, VK_DEPENDENCY_BY_REGION_BIT, 0, nullptr, 0, nullptr, 1,
+                         VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT, VK_DEPENDENCY_BY_REGION_BIT, 0, nullptr, 0, nullptr, 1,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Commit message doesn't describe why you're changing all of these. Were they already incorrect?

File: layers/core_validation.cpp [hunk 5]
```diff
@@ -6423,8 +6423,9 @@ static bool ValidateAccessMaskPipelineStage(VkAccessFlags access_mask, VkPipelin
     while (access_mask) {
         index = __builtin_ctz(access_mask);
         assert(index >= 0);
-        supported &= AccessMaskToPipeStage[index] & stage_mask;
-        access_mask >>= index;
+        supported = AccessMaskToPipeStage[index] & stage_mask;
+        if (!supported) break;  // early out
+        access_mask >>= (index + 1);
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Need to fold this in at the point you introduce this function.

File: layers/core_validation.cpp [hunk 6]
```diff
@@ -6423,8 +6423,9 @@ static bool ValidateAccessMaskPipelineStage(VkAccessFlags access_mask, VkPipelin
     while (access_mask) {
         index = __builtin_ctz(access_mask);
         assert(index >= 0);
-        supported &= AccessMaskToPipeStage[index] & stage_mask;
-        access_mask >>= index;
+        supported = AccessMaskToPipeStage[index] & stage_mask;
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: `if (!(AccessMaskToPipeStage[index] & stage_mask)) return false;`

File: tests/layer_validation_tests.cpp [hunk 7]
```diff
@@ -21140,9 +21140,8 @@ TEST_F(VkPositiveLayerTest, DepthStencilLayoutTransitionForDepthOnlyImageview) {
     imb.subresourceRange.baseArrayLayer = 0;
     imb.subresourceRange.layerCount = 0x1;
 
-    vkCmdPipelineBarrier(m_commandBuffer->handle(), VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
-                         VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT, VK_DEPENDENCY_BY_REGION_BIT, 0, nullptr, 0, nullptr, 1,
-                         &imb);
+    vkCmdPipelineBarrier(m_commandBuffer->handle(), VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: OK :)

File: layers/core_validation.cpp [hunk 8]
```diff
@@ -6362,14 +6362,96 @@ static bool ValidateRenderPassPipelineBarriers(layer_data *device_data, const ch
...
+static bool ValidateAccessMaskPipelineStage(VkAccessFlags access_mask, VkPipelineStageFlags stage_mask) {
+    // Early out if all commands set, or access_mask NULL
+    if (stage_mask & VK_PIPELINE_STAGE_ALL_COMMANDS_BIT) return true;
+    if (0 == access_mask) return false;
+    stage_mask = ExpandPipelineStageFlags(stage_mask);
+    int index = 0;
+    bool supported = true;
+    // for each of the set bits in access_mask, make sure that supporting stage mask bit(s) are set
+    while (access_mask) {
+        index = __builtin_ctz(access_mask);
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: This is a GNU special. What are you doing on MSVC?

**Comment by tobine** [comment]
Pushed updates to address @chrisforbes feedback.

**Comment by tobine** [comment]
Automated build/test is all clean now

File: layers/core_validation.cpp [hunk 9]
```diff
@@ -6362,14 +6362,98 @@ static bool ValidateRenderPassPipelineBarriers(layer_data *device_data, const ch
...
+    int index = 0;
+    bool supported = true;
+    // for each of the set bits in access_mask, make sure that supporting stage mask bit(s) are set
+    while (access_mask) {
+        index = (u_ffs(access_mask) - 1);
+        assert(index >= 0);
+        // Must have "!= 0" compare to prevent warning from MSVC
+        supported = ((AccessMaskToPipeStage[index] & stage_mask) != 0);
+        if (!supported) break;  // early out
+        access_mask >>= (index + 1);
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: This is still broken. After the first iteration, any remaining access mask bits are in the wrong place.

**Comment by tobine** [comment]
Fixed broken bit-shift loop to mask off bits of srcAccessMask so that original bit positions are preserved.

File: tests/layer_validation_tests.cpp (hunk 4 above)
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: OK, looks fine now.

**Pull request merged**